Replace isGroup logic with bundle entity based storage, Add a new Og static class

damiankloip opened the pull request:
Creating this here instead of https://github.com/Gizra/og/pull/63

The idea behind this is that we store whether or not a content entity of a specific bundle is a group or not by using the third party settings API we get on config entities. See `ThridPartySettingsInterface`. I think this makes more sense than adding the field instances that we check. That adds a lot of additional data, amongst other things.

In D7 this was done to make adding a field formatter easier. We should be able to add a field formatter where we need it quite easily without the field with the new field/entity API.

This then ties the implementation to entity types being config entities, which core does anyway for all content entities, except users.

This doesn't add a method (yet) to get all group types by entity type > bundle. Do we need that too? We could maybe do the calculation for that and cache it.

damiankloip committed:
Replace isGroup logic with bundle entity based storage, Add a new Og static class

damiankloip committed:
Save entity types in add/remove group methods, add constants for config keys

damiankloip commented on src/Og.php:
```diff
@@ -0,0 +1,106 @@
+/**
+ * A static helper class for OG.
+ */
+class Og {
```
Creating this now as the `isGroup` stuff can move to it. We should start moving stuff out of the `Drupal\og\Controller\OG` class, as controllers are meant to be for returning a response or build markup for a response (or form etc..) and could get confusing otherwise :)

amitaibu replied:
:+1:

damiankloip commented on src/Og.php:
```diff
@@ -0,0 +1,106 @@
+   *
+   * @return bool
+   */
+  public static function isGroupEntityType(ConfigEntityInterface $entity_type) {
```
This code currently relies on using config entities as the entity types. All core content entities do this, like Node, Taxonomy, Comment. However, User does not, as it just supports the single bundle. Do we support cases like the user being a group? I don't think\* that makes sense..but you never know :)

amitaibu replied:
> Do we support cases like the user being a group?

Yes we do. For example in Drupal commons we've implemented a user as group - for a facebook like functionality where user's can have friends and their own (private) content.

damiankloip replied:
Hmm OK, I thought that might be the case :) Then this is probably not the right approach.

We could actually use the same approach as the audience field, in that we have the configurable field with no field storage. Just the instances? Then at least we don't have lots of data we don't need.

We could also (I think I prefer this) have another config entity type that just stores group data. `OgGroup`? This would act as metadata for a group, so would just store the entity_type and bundle for the group. Maybe more at a later date. We could then easily use these to either cache a map of data, or just load them to find groups we need at a particular time.

amitaibu replied:
Maybe we could continue with this approach, which sounds sane, but instead cater for D8's "inconsistency" with bundles.

``` php
public static function isGroupEntityType($entity_type) {
  if ($entity_type->id() == 'user') {
    return ...
  }

  if ($entity_type instanceOf ConfigEntityInterface) {
    // ...
  }
}
```

Eventhough it's a bit unclean, it is still going to be much easier for people to grok.

damiankloip replied:
Yeah. It would keep things much simpler. We would just need to store somewhere whether the user is enabled as a group or not.

amitaibu replied:
To keep things consistent, maybe we can have a variable mark all the "group" entities. We'll just have to make sure that variable is synced with the `getThirdPartySetting` when they are available. So we'll have a "single source of truth" which is the variable.

damiankloip replied:
Yes, if we go that route, we could just not bother with the third party setting, if you want to know whether sometihng is a group or not, you ask the OG API only.

damiankloip replied:
So we just have some base OG config, and we add a list of entity type and bundles there, just like:

``` yaml
entity_type:
  - bundle
  - bundle
other_entity_type:
  - bundle
```

Thoughts?

We could then have a group storage class/manager object that controls this, it can be a service, with the base implementation just using the simple configuration described above.

amitaibu replied:
Yes, that list sounds right
On Sep 19, 2015 12:10 PM, "Damian Lee" notifications@github.com wrote:

> In src/Og.php https://github.com/amitaibu/og/pull/2#discussion_r39918186
> :
> 
> > +
> > -      return $is_group;
> > -    }
> >   +
> > -    static::$groupEntityCache[$entity_type_id][$entity_bundle] = FALSE;
> > -    return FALSE;
> > -  }
> >   +
> > -  /**
> > -   \* Returns whether an entity type instance is a group type.
> > -   *
> > -   \* @param \Drupal\Core\Config\Entity\ConfigEntityInterface $entity_type
> > -   *
> > -   \* @return bool
> > -   */
> > -  public static function isGroupEntityType(ConfigEntityInterface $entity_type) {
> 
> So we just have some base OG config, and we add a list of entity type and
> bundles there, just like:
> 
> entity_type:
> - bundle
> - bundleother_entity_type:
> - bundle
> 
> Thoughts?
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/amitaibu/og/pull/2/files#r39918186.

damiankloip committed:
Remove OG_GROUP_FIELD constant

damiankloip committed:
Replace group storage config with og.settings config file

damiankloip committed:
Fix config schema

damiankloip commented:
@amitaibu re implemented just storing in configuration settings. This is a much simpler approach.

damiankloip commented:
@RoySegall ^^

amitaibu commented:
@damiankloip ready for review?

damiankloip commented:
Nearly, I have created a group manager. I was going to keep the convenience methods on the Og class too and just call out to the group manager, like :

``` php
// In Og static class...

public static function entityIsGroup(EntityInterface $entity) {
  return \Drupal::service('og.group.manager')->entityIsGroup($entity);
}
```

Thoughts? That seems better (esp. at the moment) than always having to get the group manager service, Just using:

``` php
Og::entityIsGroup($entity);
```

damiankloip committed:
Add bundle lookup method, use group manager in Og static class

damiankloip commented:
ping @amitaibu @RoySegall , what do you guys think of this now? I have added a group manager, which makes sense to control all storage that determines what a group is. I have added the Og static class to replace the Controller/OG class. This class calls out to the group manager but just allows easy calling with static methods `Og::isGroup($entity)` etc..

I also added a helper to look up whether a group is a bundle by passing in the actual bundle entity. I assume that's useful? Do you think there are any other methods I should add at this stage?

damiankloip commented:
E.g. maybe Og::addgroup should accept an entity instead of Id and bundle? and then pass that to the group manager? It might be a good idea to keep all entity related logic out of the group manager, and KISS. Then it just accepts entity ID and bundle ID as params all the time, as that's all it needs for the mapping? But, on the flipside of that - if we inject this into place, we might want those additional conveniences.

RoySegall commented:
From an overall look it's look good but until i'll test by my self or if we have test to see the API I don't have a concrete answer.

Another question - now that we have the group manager service it's mean that we have it as a dependency injection? Because i can see in a lot places that there are services calls(like the entity manager) and also there are some controllers which have the entity manager as dependency injection(maybe this is the best practice for now).

damiankloip commented:
Yes, once we start getting other proper services/controllers we can inject this directly. I think having the Og:: methods is still good. especially for now. There will still likely be procedural code calling this too. I can work on adding some tests.

Maybe just starting with unit tests for the group manager is a good idea. We should have a follow up for more integration type tests I think, as we can now use the custom_storage property on our fields. yeah!!

RoySegall commented:
Yes, this is where i'm heading. We can keep this is it is since we in a static methods and injection services can be kind of redundant. But we will use it as dependency injection(let's use DI from now) for og ui forms for example.

I think simpletest could do the trick but phpunit is good direction because I was think on phpunit for testing the code and behat for testing UI stuff.

amitaibu commented:
> But, on the flipside of that - if we inject this into place, we might want those additional conveniences.

In OG7 we had this in a few functions that used to get both ID and entity. It has it's advantages, but it's also a bit ambiguous. We can start with getting just the ID, and if we see that we have to extract it too many times we can reconsider.

After that, let's merge :)

damiankloip commented:
ok, sounds good, I can remove the entity\* methods. They are just small wrappers. The group manager is essentially a key value store. I will write some tests today for that service, then we should be looking good! Cheers guys

amitaibu commented:
:+1:

damiankloip committed:
Add unit test for OgGroupManager

damiankloip committed:
Remove entityIsGroup and entityBundleIsGroup from GroupManager

damiankloip commented:
@amitaibu Added a unit test for `OgGroupManager` and remove the entity related methods.

I think I should just rename to `GroupManager`, as we are already in the og namespace. So > `Drupal\og\GroupManager`. Thoughts?

damiankloip committed:
Remove entity based group methods from Og static class

amitaibu commented on tests/src/Unit/OgGroupManagerTest.php:
```diff
@@ -0,0 +1,227 @@
+   * {@inheritdoc}
+   */
+  public function setUp() {
+    $this->configProphecy = $this->prophesize('Drupal\Core\Config\Config');
```
oh man, I should learn those new stuff, I have no idea what it's doing :/

amitaibu replied:
but thanks for knowing it yourself :)

damiankloip replied:
Ha. Yeah! It's pretty useful.

amitaibu commented:
> I think I should just rename to GroupManager, as we are already in the og namespace. So > Drupal\og\GroupManager. Thoughts?

It might evolve to become even just `Manager` once we port more thing, but for now sure, let's change it to `GroupManager`

amitaibu commented on src/OgGroupManager.php:
```diff
@@ -0,0 +1,137 @@
+   * @return bool
+   */
+  public function isGroup($entity_type_id, $bundle) {
+    if (!isset($this->groupMap)) {
```
this IF part that is duplicated bugs me a bit, it implies we don't have full control over its state. can't we move it to the constructor?

damiankloip committed:
Rename OgGroupManager > GroupManager

damiankloip commented:
Rename complete.

amitaibu commented:
The other issues (the repeating IF) can be dealt with follow up.

amitaibu commented:
Thanks!
